File: frontend/src/types/awards.ts
// Award types - 球员荣誉/奖项类型

export type AwardType =
  | 'match_mvp'
  | 'league_team_of_season'
  | 'league_best_fw' | 'league_best_mf' | 'league_best_df' | 'league_best_gk'
  | 'league_golden_boot' | 'league_playmaker' | 'league_golden_glove' | 'league_golden_wall'
  | 'cup_golden_boot' | 'cup_playmaker' | 'cup_golden_glove' | 'cup_golden_wall'
  | 'season_best_player'
  | 'season_best_fw' | 'season_best_mf' | 'season_best_df' | 'season_best_gk'
  | 'season_golden_boot' | 'season_playmaker' | 'season_golden_glove' | 'season_golden_wall'

export type AwardLevel = 'match' | 'league' | 'cup' | 'season'

export interface AwardMetadata {
  rating?: number
  matches?: number
  goals?: number
  assists?: number
  clean_sheets?: number
  tackles?: number
  interceptions?: number
  saves?: number
  championships?: number
  mvp_count?: number
  primary_value?: number
  position_rank?: number
  team?: string
  opponent?: string
  match_result?: string
  cup_name?: string
}

export interface PlayerAward {
  id: string
  player_id: string
  player_name?: string
  player_avatar_url?: string
  player_position?: string
  season_id: string
  season_number: number
  award_type: AwardType
  award_level: AwardLevel
  league_id?: string
  league_name?: string
  cup_id?: string
  cup_name?: string
  fixture_id?: string
  position?: 'FW' | 'MF' | 'DF' | 'GK'
  metadata?: AwardMetadata
  description?: string
  created_at: string
}

export interface PlayerAwardSummary {
  total_awards: number
  mvp_count: number
  team_of_season_count: number
  best_position_count: number
  golden_boot_count: number
  playmaker_count: number
  golden_glove_count: number
  golden_wall_count: number
  season_best_player_count: number
}

export interface SeasonAwards {
  season_id: string
  season_number: number
  best_player?: PlayerAward
  best_fw?: PlayerAward
  best_mf?: PlayerAward
  best_df?: PlayerAward
  best_gk?: PlayerAward
  golden_boot?: PlayerAward
  playmaker?: PlayerAward
  golden_glove?: PlayerAward
  golden_wall?: PlayerAward
}

export interface LeagueAwards {
  league_id: string
  season_id: string
  season_number: number
  team_of_season: PlayerAward[]
  best_fw?: PlayerAward
  best_mf?: PlayerAward
  best_df?: PlayerAward
  best_gk?: PlayerAward
  golden_boot?: PlayerAward
  playmaker?: PlayerAward
  golden_glove?: PlayerAward
  golden_wall?: PlayerAward
}

export interface CupAwards {
  cup_id: string
  season_id: string
  season_number: number
  golden_boot?: PlayerAward
  playmaker?: PlayerAward
  golden_glove?: PlayerAward
  golden_wall?: PlayerAward
}

export const AWARD_LABELS: Record<AwardType, string> = {
  match_mvp: '本场最佳',
  league_team_of_season: '联赛最佳阵容',
  league_best_fw: '联赛最佳前锋',
  league_best_mf: '联赛最佳中场',
  league_best_df: '联赛最佳后卫',
  league_best_gk: '联赛最佳门将',
  league_golden_boot: '联赛金靴奖',
  league_playmaker: '联赛助攻王',
  league_golden_glove: '联赛金手套奖',
  league_golden_wall: '联赛金墙奖',
  cup_golden_boot: '杯赛金靴奖',
  cup_playmaker: '杯赛助攻王',
  cup_golden_glove: '杯赛金手套奖',
  cup_golden_wall: '杯赛金墙奖',
  season_best_player: '闪电足球先生',
  season_best_fw: '年度最佳前锋',
  season_best_mf: '年度最佳中场',
  season_best_df: '年度最佳后卫',
  season_best_gk: '年度最佳门将',
  season_golden_boot: '赛季金靴奖',
  season_playmaker: '赛季助攻王',
  season_golden_glove: '赛季金手套奖',
  season_golden_wall: '赛季金墙奖',
}

export const AWARD_ICONS: Record<AwardType, string> = {
  match_mvp: '⚽',
  league_team_of_season: '⭐',
  league_best_fw: '🔥',
  league_best_mf: '🔥',
  league_best_df: '🛡️',
  league_best_gk: '🧤',
  league_golden_boot: '🥾',
  league_playmaker: '👟',
  league_golden_glove: '🧤',
  league_golden_wall: '🧱',
  cup_golden_boot: '🥾',
  cup_playmaker: '👟',
  cup_golden_glove: '🧤',
  cup_golden_wall: '🧱',
  season_best_player: '👑',
  season_best_fw: '🔥',
  season_best_mf: '🔥',
  season_best_df: '🛡️',
  season_best_gk: '🧤',
  season_golden_boot: '🥾',
  season_playmaker: '👟',
  season_golden_glove: '🧤',
  season_golden_wall: '🧱',
}

export const AWARD_COLORS: Record<AwardType, string> = {
  match_mvp: 'text-amber-400',
  league_team_of_season: 'text-cyan-400',
  league_best_fw: 'text-red-400',
  league_best_mf: 'text-emerald-400',
  league_best_df: 'text-blue-400',
  league_best_gk: 'text-amber-400',
  league_golden_boot: 'text-yellow-400',
  league_playmaker: 'text-green-400',
  league_golden_glove: 'text-sky-400',
  league_golden_wall: 'text-orange-400',
  cup_golden_boot: 'text-yellow-400',
  cup_playmaker: 'text-green-400',
  cup_golden_glove: 'text-sky-400',
  cup_golden_wall: 'text-orange-400',
  season_best_player: 'text-[#C6F135]',
  season_best_fw: 'text-red-400',
  season_best_mf: 'text-emerald-400',
  season_best_df: 'text-blue-400',
  season_best_gk: 'text-amber-400',
  season_golden_boot: 'text-yellow-400',
  season_playmaker: 'text-green-400',
  season_golden_glove: 'text-sky-400',
  season_golden_wall: 'text-orange-400',
}

File: backend/app/routers/awards.py
"""
Awards router - 球员荣誉/奖项 API
"""
from typing import Optional, List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db
from app.schemas.base import ResponseSchema
from app.schemas.award import (
    PlayerAwardResponse,
    PlayerAwardSummary,
    SeasonAwardsResponse,
    LeagueAwardsResponse,
    CupAwardsResponse,
)
from app.services.award_service import AwardService
from app.models.player_award import PlayerAward, AwardType, AwardLevel
from app.models.player import Player
from app.models.league import League
from app.models.season import Season
from app.models.season import CupCompetition
from sqlalchemy import select

router = APIRouter(prefix="/awards", tags=["awards"])


def _award_to_response(award: PlayerAward) -> PlayerAwardResponse:
    """将 PlayerAward 模型转换为响应 schema"""
    return PlayerAwardResponse(
        id=award.id,
        player_id=award.player_id,
        player_name=award.player.name if award.player else None,
        player_avatar_url=award.player.avatar_url if award.player else None,
        player_position=award.player.position.value if award.player and award.player.position else None,
        season_id=award.season_id,
        season_number=award.season_number,
        award_type=award.award_type.value,
        award_level=award.award_level.value,
        league_id=award.league_id,
        league_name=award.league.name if award.league else None,
        cup_id=award.cup_id,
        cup_name=award.cup.name if award.cup else None,
        fixture_id=award.fixture_id,
        position=award.position,
        metadata=award.award_metadata,
        description=award.description,
        created_at=award.created_at,
    )


# ===== 手动触发接口（调试用） =====

@router.post("/match-mvp/{fixture_id}", response_model=ResponseSchema[Optional[PlayerAwardResponse]])
async def trigger_match_mvp(fixture_id: str, db: AsyncSession = Depends(get_db)):
    """手动触发单场 MVP 评选"""
    from app.models.match_result import MatchResult
    result = await db.get(MatchResult, fixture_id)
    if not result:
        raise HTTPException(status_code=404, detail="比赛结果不存在")

    award = await AwardService.award_match_mvp(fixture_id, result, db)
    await db.commit()
    if award:
        # 刷新关联对象
        await db.refresh(award, ["player", "season", "league", "cup"])
    return ResponseSchema(
        success=award is not None,
        data=_award_to_response(award) if award else None,
    )


@router.post("/league/{league_id}/season-end", response_model=ResponseSchema[List[PlayerAwardResponse]])
async def trigger_league_awards(league_id: str, season_id: str, db: AsyncSession = Depends(get_db)):
    """手动触发联赛奖项评选"""
    awards = await AwardService.award_league_end_of_season(league_id, season_id, db)
    await db.commit()
    for a in awards:
        await db.refresh(a, ["player", "league"])
    return ResponseSchema(
        success=True,
        data=[_award_to_response(a) for a in awards],
    )


@router.post("/cup/{cup_id}/end", response_model=ResponseSchema[List[PlayerAwardResponse]])
async def trigger_cup_awards(cup_id: str, season_id: str, db: AsyncSession = Depends(get_db)):
    """手动触发杯赛奖项评选"""
    awards = await AwardService.award_cup_end(cup_id, season_id, db)
    await db.commit()
    for a in awards:
        await db.refresh(a, ["player", "cup"])
    return ResponseSchema(
        success=True,
        data=[_award_to_response(a) for a in awards],
    )


@router.post("/season-end/{season_id}", response_model=ResponseSchema[List[PlayerAwardResponse]])
async def trigger_season_awards(season_id: str, db: AsyncSession = Depends(get_db)):
    """手动触发赛季大奖评选"""
    awards = await AwardService.award_season_end(season_id, db)
    await db.commit()
    for a in awards:
        await db.refresh(a, ["player"])
    return ResponseSchema(
        success=True,
        data=[_award_to_response(a) for a in awards],
    )


# ===== 查询接口 =====

@router.get("/player/{player_id}", response_model=ResponseSchema[List[PlayerAwardResponse]])
async def get_player_awards(player_id: str, db: AsyncSession = Depends(get_db)):
    """获取球员荣誉列表"""
    awards = await AwardService.get_player_awards(player_id, db)
    return ResponseSchema(
        success=True,
        data=[_award_to_response(a) for a in awards],
    )


@router.get("/player/{player_id}/summary", response_model=ResponseSchema[PlayerAwardSummary])
async def get_player_award_summary(player_id: str, db: AsyncSession = Depends(get_db)):
    """获取球员荣誉统计摘要"""
    summary = await AwardService.get_player_award_summary(player_id, db)
    return ResponseSchema(
        success=True,
        data=PlayerAwardSummary(**summary),
    )


@router.get("/season/{season_id}", response_model=ResponseSchema[SeasonAwardsResponse])
async def get_season_awards(season_id: str, db: AsyncSession = Depends(get_db)):
    """获取某赛季全部大奖"""
    season = await db.get(Season, season_id)
    if not season:
        raise HTTPException(status_code=404, detail="赛季不存在")

    grouped = await AwardService.get_season_awards(season_id, db)

    def _first(awards: List[PlayerAward]) -> Optional[PlayerAwardResponse]:
        if not awards:
            return None
        return _award_to_response(awards[0])

    for awards in grouped.values():
        for a in awards:
            await db.refresh(a, ["player"])

    return ResponseSchema(
        success=True,
        data=SeasonAwardsResponse(
            season_id=season_id,
            season_number=season.season_number,
            best_player=_first(grouped.get(AwardType.SEASON_BEST_PLAYER, [])),
            best_fw=_first(grouped.get(AwardType.SEASON_BEST_FW, [])),
            best_mf=_first(grouped.get(AwardType.SEASON_BEST_MF, [])),
            best_df=_first(grouped.get(AwardType.SEASON_BEST_DF, [])),
            best_gk=_first(grouped.get(AwardType.SEASON_BEST_GK, [])),
            golden_boot=_first(grouped.get(AwardType.SEASON_GOLDEN_BOOT, [])),
            playmaker=_first(grouped.get(AwardType.SEASON_PLAYMAKER, [])),
            golden_glove=_first(grouped.get(AwardType.SEASON_GOLDEN_GLOVE, [])),
            golden_wall=_first(grouped.get(AwardType.SEASON_GOLDEN_WALL, [])),
        ),
    )


@router.get("/league/{league_id}", response_model=ResponseSchema[LeagueAwardsResponse])
async def get_league_awards(
    league_id: str,
    season_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取某联赛某赛季奖项"""
    league = await db.get(League, league_id)
    if not league:
        raise HTTPException(status_code=404, detail="联赛不存在")

    awards = await AwardService.get_league_awards(league_id, season_id, db)
    for a in awards:
        await db.refresh(a, ["player"])

    def _first_of_type(atype: AwardType) -> Optional[PlayerAwardResponse]:
        items = [a for a in awards if a.award_type == atype]
        return _award_to_response(items[0]) if items else None

    season = await db.get(Season, season_id)
    season_number = season.season_number if season else 0

    return ResponseSchema(
        success=True,
        data=LeagueAwardsResponse(
            league_id=league_id,
            season_id=season_id,
            season_number=season_number,
            team_of_season=[_award_to_response(a) for a in awards if a.award_type == AwardType.LEAGUE_TEAM_OF_SEASON],
            best_fw=_first_of_type(AwardType.LEAGUE_BEST_FW),
            best_mf=_first_of_type(AwardType.LEAGUE_BEST_MF),
            best_df=_first_of_type(AwardType.LEAGUE_BEST_DF),
            best_gk=_first_of_type(AwardType.LEAGUE_BEST_GK),
            golden_boot=_first_of_type(AwardType.LEAGUE_GOLDEN_BOOT),
            playmaker=_first_of_type(AwardType.LEAGUE_PLAYMAKER),
            golden_glove=_first_of_type(AwardType.LEAGUE_GOLDEN_GLOVE),
            golden_wall=_first_of_type(AwardType.LEAGUE_GOLDEN_WALL),
        ),
    )


@router.get("/cup/{cup_id}", response_model=ResponseSchema[CupAwardsResponse])
async def get_cup_awards(
    cup_id: str,
    season_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取某杯赛某赛季奖项"""
    cup = await db.get(CupCompetition, cup_id)
    if not cup:
        raise HTTPException(status_code=404, detail="杯赛不存在")

    awards = await AwardService.get_cup_awards(cup_id, season_id, db)
    for a in awards:
        await db.refresh(a, ["player"])

    def _first_of_type(atype: AwardType) -> Optional[PlayerAwardResponse]:
        items = [a for a in awards if a.award_type == atype]
        return _award_to_response(items[0]) if items else None

    season = await db.get(Season, season_id)
    season_number = season.season_number if season else 0

    return ResponseSchema(
        success=True,
        data=CupAwardsResponse(
            cup_id=cup_id,
            season_id=season_id,
            season_number=season_number,
            golden_boot=_first_of_type(AwardType.CUP_GOLDEN_BOOT),
            playmaker=_first_of_type(AwardType.CUP_PLAYMAKER),
            golden_glove=_first_of_type(AwardType.CUP_GOLDEN_GLOVE),
            golden_wall=_first_of_type(AwardType.CUP_GOLDEN_WALL),
        ),
    )


@router.get("/season/{season_id}/leagues", response_model=ResponseSchema[List[LeagueAwardsResponse]])
async def get_all_league_awards_for_season(
    season_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取某赛季所有联赛的奖项"""
    season = await db.get(Season, season_id)
    if not season:
        raise HTTPException(status_code=404, detail="赛季不存在")

    # 获取所有联赛
    leagues_result = await db.execute(select(League))
    leagues = leagues_result.scalars().all()

    results: List[LeagueAwardsResponse] = []
    for league in leagues:
        awards = await AwardService.get_league_awards(league.id, season_id, db)
        for a in awards:
            await db.refresh(a, ["player"])

        def _first_of_type(atype: AwardType) -> Optional[PlayerAwardResponse]:
            items = [a for a in awards if a.award_type == atype]
            return _award_to_response(items[0]) if items else None

        results.append(LeagueAwardsResponse(
            league_id=league.id,
            season_id=season_id,
            season_number=season.season_number,
            team_of_season=[_award_to_response(a) for a in awards if a.award_type == AwardType.LEAGUE_TEAM_OF_SEASON],
            best_fw=_first_of_type(AwardType.LEAGUE_BEST_FW),
            best_mf=_first_of_type(AwardType.LEAGUE_BEST_MF),
            best_df=_first_of_type(AwardType.LEAGUE_BEST_DF),
            best_gk=_first_of_type(AwardType.LEAGUE_BEST_GK),
            golden_boot=_first_of_type(AwardType.LEAGUE_GOLDEN_BOOT),
            playmaker=_first_of_type(AwardType.LEAGUE_PLAYMAKER),
            golden_glove=_first_of_type(AwardType.LEAGUE_GOLDEN_GLOVE),
            golden_wall=_first_of_type(AwardType.LEAGUE_GOLDEN_WALL),
        ))

    return ResponseSchema(success=True, data=results)
